docs/docs/architecture/client_broker_protocol.md

```diff
@@ -0,0 +1,1223 @@
+`AckCallback` and then invoke it to deliver the result to the application.
+
+### Receiving Messages (Consumers)
+
+An application which opens a queue as a consumer (i.e., with *READ* intent)
+will receive PUSH messages from the broker anytime a new message is posted on
+the queue.  Like ACK messages, PUSH messages can be delivered in a batch in a
+PUSH event, and the SDK must be ready to handle it.
+
+#### Consume API
+
+As far as handing over PUSH messages to the application is concerned, it can be
+done by asking for a callback from the application when it opens the queue, and
+dispatching PUSH messages by invoking that callback.  There are various ways to
+ask applications for this callback, but the most obvious one could be in the
+`openQueue` API (or a flavor of it).  For example, Java SDK has a
+[`PushMessageHandler`](https://github.com/bloomberg/blazingmq-sdk-java/blob/main/bmq-sdk/src/main/java/com/bloomberg/bmq/PushMessageHandler.java)
+interface, which consumer application should specify when
+[creating](https://github.com/bloomberg/blazingmq-sdk-java/blob/ed0eb848f5ac2897cfe32cc481575d689a4cb1c2/bmq-examples/src/main/java/com/bloomberg/bmq/examples/SimpleConsumer.java#L117C42-L117C42)
+a queue.
+
+Since multiple PUSH messages can appear in a PUSH event, the SDK implementation
+must be able to handle that, and it may need to dispatch PUSH messages one by
+one in the callback if the callback's signature does not support handing over
+multiple PUSH messages to the application in one go.
+
+#### Consume API Implementation
+
+Wire layout of a PUSH event was described in a [previous](#push-event) section.
+As far as implemention is concerned, SDK logic should keep some of these things
+in mind when receiving and dispatching a PUSH event:
+
+- Ensure that the queue is still open, and not closed or being closed by the
+  application.
+
+- Ensure that the PUSH event parsing passes basic sanity checks e.g., length of
+  the packet is equal to the length specified in the `EventHeader`, etc
+
+- Ensure that the `queueId` field specified in the `PushHeader` is valid and
+  known to the SDK
+
+### Confirming Messages (Consumers)
+
+After opening a queue, a consumer application will receive messages which are
+being posted on the queue.  Some details about handling PUSH messages were
+described in the previous section.  One the consumer application processes a
+PUSH message by executing its business logic, it needs to indicate to the
+BlazingMQ backend that message has been processed successfully and can be
+removed from the queue.  This is done by the consumer application sending a
+CONFIRM message.
+
+#### ConfirmMessage API
+
+Similar to the **Post API**, C++ and Java SDKs support batching of CONFIRM
+messages i.e., consumer application can send a batch of CONFIRM messages in one
+shot instead of sending them one by one.  Batching can have an improved impact
+on the performance.  Additionally, in both SDKs, the `confirmMessage()` API is
```

Comment:
```suggestion
shot instead of sending them one by one.  Batching can improve performance.
Additionally, in both SDKs, the `confirmMessage()` API is
```
or
```suggestion
shot instead of sending them one by one.  Batching can have a positive impact
on the performance.  Additionally, in both SDKs, the `confirmMessage()` API is
```